src/persist/src/file.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! In-memory implementations for testing and benchmarking.
+
+use std::collections::{HashMap, HashSet};
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Seek, SeekFrom, Write as StdWrite};
+use std::sync::{Arc, Mutex};
+
+use crate::error::Error;
+use crate::mem::MemSnapshot;
+use crate::persister::{Meta, Persister, Write};
+use crate::{Id, Token};
+
+/// A naive implementation of [Write] and [Meta] backed by a file.
+#[derive(Clone, Debug)]
+pub struct FileStream {
+    dataz: Arc<Mutex<File>>,
+}
+
+impl FileStream {
+    /// Create a new FileStream
+    pub fn new(name: String) -> Self {
+        // TODO: could not get truncate to work here.
+        let file = OpenOptions::new()
+            .write(true)
+            .append(true)
+            .create(true)
+            .read(true)
+            .open(name)
+            .expect("failed to create file");
+        FileStream {
+            dataz: Arc::new(Mutex::new(file)),
+        }
+    }
+}
+
+impl Write for FileStream {
+    fn write_sync(&mut self, updates: &[((String, String), u64, isize)]) -> Result<(), Error> {
+        let mut buf = Vec::new();
```

Comment:
Put this on the FileStream and truncate it to share the alloc between calls (premature non-profile drivin optimization, yes, but an easy and clear one)